File: kevin-bacon/traverse.cpp
//  This is also written in c++ because of the size of the data involved. \
	Without any optimization or parallelization of input files, they take \
	up over 5GB of RAM. In Python they would undoubtedly take up more.    \
	This is an issue on my laptop, as it has only 4GB of RAM and 7GB of swap.

// imports
#include <unordered_set>
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <stack>
using namespace std;

//  datastructure to hold a node in adjacency list - it stores an integer \
	index, number of adjacent vertices, and an array of adjacent node     \
	indexes (stored as a heap array)
typedef struct{
	int id;
	int n_adj;
	int *adj;
	bool init = 0;
} node;

//  datastructure to hold node in traversal (holds depth, index, and which \
	node it branched from
typedef struct{
	int id;
	int depth;
	int from;
	bool is_act = 1;
} traverse_node;

// utility to draw progress bars
long long prc=100;
void progress(long long p) {
	if(p==100||p>prc||(p==0&&prc>0)) prc = p;
	else return;
	cout << "\r[";
	for(long long i=0;i<100;i+=2){
		if(i<p-1) cout << '=';
		else if(i==p-1) cout << '-';
		else cout << ' ';
	}
	cout << "] (" << p << "%)  ";
	fflush(stdout);
	if(p>=100) cout << '\n';
}

// global variables
const long long name_len = 13205098,
	            akas_len = 38592396;

// an, vn - {actor, movie} namelist pointers
string *an, *vn;
// al, vl - {actor, movie} adjacency list pointers
node *al, *vl;

// utility to create new nodes in traversal tree
traverse_node actor_node_at(int id, int depth, int from) {
	traverse_node tmp;
	tmp.id = id;
	tmp.depth = depth;
	tmp.from = from;
	return tmp;
}

traverse_node movie_node_at(int id, int depth, int from) {
	traverse_node tmp;
	tmp.id = id;
	tmp.is_act = 0;
	tmp.depth = depth;
	tmp.from = from;
	return tmp;
}

// overloads
traverse_node actor_node_at(int id) {
	return actor_node_at(id, 0, -1);
}

traverse_node movie_node_at(int id) {
	return movie_node_at(id, 0, -1);
}

// asks user for actor name until "exit" is entered
void search(bool first) {
	// ask for input, and display tip on first run
	if(first) cout << "Enter the name of an actor, or 'exit' to exit:\n";
	cout << "> ";
	string actor; getline(cin, actor);
	
	// exit if exit is entered
	if(actor == "exit") return;
	
	// find actor in actor namelist
	cout << "Searching for " << actor << " in namelist\n";
	
	int actor_id = -1;
	
	for(int i=0;i<name_len;i++){
		if(an[i] == actor){
			actor_id = i;
			break;
		}
		progress(i*100/name_len);
	}
	
	//  actor_id is by default -1, so if it is still -1, we have not found \
		the actor, so we exit this run of the method
	if(actor_id+1)
		cout << "\nFound " << actor << " at index " << actor_id << '\n';
	else{
		cout << "\nCouldn't find " << actor << " in namelist\n\n";
		search(0);
		// avoid complications
		return;
	}
	
	//  I wasn't able to use DFS due to the extreme size of the graph, \
		so BFS was necessary
	cout << "Building search queue\n";
	
	//  We actually want to keep every node that ever has entered the queue, \
		in order to backtrace to see what path we took (DFS doesn't have     \
		this issue, but a shortest path was necessary. I used a vector.
	vector<traverse_node> search_queue;
	int front_index = 0;
	
	// avoid repeating paths already traversed
	unordered_set<int> actors_visited;
	unordered_set<int> movies_visited;
	
	// initialize queue
	search_queue.push_back(actor_node_at(actor_id));
	actors_visited.insert(actor_id);
	
	cout << "Moving through graph\n";
	
	// keep track of "front node"
	int iter = 0, depth = 0;
	traverse_node front = search_queue[0];
	
	// we haven't found him yet!
	bool found_kevin_bacon = 0;
	// well, unless we started with him..
	if(actor == "Kevin Bacon") found_kevin_bacon = 1;
	
	// index of Kevin Bacon is 101
	while(!found_kevin_bacon && front_index < search_queue.size()){
		// update front node
		front = search_queue[front_index];
		front_index++;
		
		/*
		cout << front.id << " | ";
		if(front.is_act) cout << an[front.id];
		else cout << vn[front.id];
		cout << "\n\n";
		*/
		
		bool finished = 0;
		
		// bistate graph - nodes alternate between actor and movie
		if(front.is_act){
			// loop through all adjacent and append to queue
			if(al[front.id].init) for(int i=0;i<al[front.id].n_adj;i++){
				if(movies_visited.find(al[front.id].adj[i])
				!=movies_visited.end()) continue;
				movies_visited.insert(al[front.id].adj[i]);
				search_queue.push_back(
					movie_node_at(
						al[front.id].adj[i], front.depth+1, iter
					)
				);
			}
		}else{
			if(vl[front.id].init) for(int i=0;i<vl[front.id].n_adj;i++){
				if(actors_visited.find(vl[front.id].adj[i])
				!=actors_visited.end()) continue;
				actors_visited.insert(vl[front.id].adj[i]);
				search_queue.push_back(
					actor_node_at(
						vl[front.id].adj[i], front.depth+1, iter
					)
				);
				// found him!
				if(vl[front.id].adj[i] == 101){
					front = search_queue[search_queue.size()-1];
					found_kevin_bacon = 1;
					finished = 1;
					break;
				}
			}
		}

		if(finished) break;
		
		depth = front.depth;
		
		// update depth
		if(iter%13==0){
			cout << "\rDepth " << front.depth << " (" << iter << " iterations)";
			fflush(stdout);
		}
		
		iter++;
	}
	
	if(!found_kevin_bacon){
		cout << "\nCouldn't find a path from " << actor
			 << " to Kevin Bacon.\n\n";
		
		// rerun program
		search(0);
		return;
	}
	
	cout << "\nFound Kevin Bacon!\n";
	
	// because we need to backtrace, we use a stack
	stack<traverse_node> path;
	
	//  while front is not the root node (from -1), backtrace to its \
		root node
	while(front.from+1 && depth--> -3){
		path.push(front);
		
		front = search_queue[front.from];
	}
	
	if(depth < -2) cout << "An error might have occured.\n";
	
	// print out the path
	cout << "Path:\n";
	
	cout << actor;
	
	// print everything inside stack
	while(!path.empty()){
		traverse_node node = path.top();
		path.pop();
		
		if(node.is_act){
			// get name from namelist and print
			cout << an[node.id];
			if(node.id == 101) break;
		}else{
			cout << " --- (" << vn[node.id] <<  ") --> ";
		}
	}
	
	cout << "\n\n";
	
	// rerun
	search(0);
}

int main() {
	// initialze file streams
	cout << "Opening file streams\n";

	ifstream act_name("actors_name.list");
	ifstream act_adj("actors_adj.list");
	ifstream mov_name("movies_name.list");
	ifstream mov_adj("movies_adj.list");
	
	progress(100);
	
	// create arrays to store {actor, movie} name, adjlist
	cout << "Creating datastructures\n";
	
	progress(0);
	
	string *act_namelist = new string[name_len];
	
	// update global pointers
	an = act_namelist;
	
	node *act_adjlist = new node[name_len];
	
	// fill node ids
	for(int i=0;i<name_len;i++){
		act_adjlist[i].id = i;
		progress(i*100/(name_len+akas_len));
	}
	
	al = act_adjlist;
	
	string *mov_namelist = new string[akas_len];
	
	vn = mov_namelist;
	
	node *mov_adjlist = new node[akas_len];
	
	for(int i=0;i<akas_len;i++){
		mov_adjlist[i].id = i;
		progress((name_len+i+1)*100/(name_len+akas_len));
	}
	
	vl = mov_adjlist;
	
	// now we need to read from the files and process into our arrays
	cout << "Building adjacency lists\n";
	
	for(int i=0;i<name_len;i++){
		//  our adjacency lists contain the number of adjacent vertices \
			and then all of the vertices; we must then read in all of them
		int n_adjacent; act_adj >> n_adjacent;
		// update properties of object
		act_adjlist[i].n_adj = n_adjacent;
		// initialize heap array
		act_adjlist[i].adj = new int[n_adjacent];
		for(int j=0;j<n_adjacent;j++){
			act_adj >> act_adjlist[i].adj[j];
		}
		act_adjlist[i].init = 1;
		progress(i*100/(name_len+akas_len));
	}
	
	for(int i=0;i<akas_len;i++){
		int n_adjacent; mov_adj >> n_adjacent;
		mov_adjlist[i].n_adj = n_adjacent;
		mov_adjlist[i].adj = new int[n_adjacent];
		for(int j=0;j<n_adjacent;j++){
			mov_adj >> mov_adjlist[i].adj[j];
		}
		mov_adjlist[i].init = 1;
		progress((i+name_len+1)*100/(name_len+akas_len));
	}
	
	cout << "Reading names\n";
	
	for(int i=0;i<name_len;i++){
		getline(act_name, act_namelist[i]);
		progress(i*100/(name_len+akas_len));
	}
	
	for(int i=0;i<akas_len;i++){
		getline(mov_name, mov_namelist[i]);
		progress((i+name_len+1)*100/(name_len+akas_len));
	}
	
	cout << "Closing file streams\n";
	
	act_adj.close();
	act_name.close();
	mov_adj.close();
	mov_name.close();
	
	progress(100);
	
	cout << "Ready\n";
	
	// this is the first iteration so we run with first = true
	search(1);
	
	// clean up arrays
	cout << "Deleting datastructures\n";
	
	progress(0);
	
	delete[] act_namelist;
	
	progress(25);
	
	//  because each node has a heap array attached, we must explicitly \
		delete it, and then delete the outer array
	for(int i=0;i<name_len;i++){
		delete[] act_adjlist[i].adj;
		progress(25+i*25/name_len);
	}
	
	delete[] act_adjlist;
	
	progress(50);
	
	delete[] mov_namelist;
	
	progress(75);
	
	for(int i=0;i<akas_len;i++){
		delete[] mov_adjlist[i].adj;
		progress(75+i*25/akas_len);
	}
	
	delete[] mov_adjlist;
	
	progress(100);
	
	cout << "Exiting\n";
}

File: kevin-bacon/traverse.py
"""
	Disclaimer: Unless you have *a lot* of memory, this program
	probably won't run. Likely it'll hang on 'Opening files', possibly
	to the point of noticeably slowing down the rest of your computer.
	It could even cause a kpanic, and unexpectedly reboot. I don't know.
	
	I strongly suggest using the c++ version of this program as it is
	at least 100x faster and uses less memory. Also, its code is more
	interesting - my method of data representation is somewhat unorthodox,
	and it can't be appreciated in as abstracted of a language as Python is.
	
	I wasn't able to easily trim down the dataset - the average order of
	each vertex was high enough that data trimming either caused lots of
	completely disconnected regions, or big clumps of people who were all
	mutually connected. And manufacturing an equivalent dataset defeats
	the purpose of the excercise.
	
	However - if you *really really* want to run this, be my guest - just
	make sure to allocate a lot of swap space, configure your system to let
	python use several gigabytes of RAM, and close out of any running apps
	with unsaved work.
	
	Maybe say a prayer or two.
"""

class Node:
	def __init__(self, i):
		self.id = i
		self.adj = None
		self.init = False
	
	def fill(self, adj):
		self.adj = adj;
		self.init = True

class TraverseNode:
	def __init__(self, i, prev, is_actor):
		self.id = i
		self.prev = prev
		self.depth = prev.depth+1 if prev else 0
		self.is_actor = is_actor

def progress(p):
	print(f"{round(p/10)*10}% complete")

print("Initializing datastructures")

progress(0)

actor_namelist, movie_namelist = [], []
name_len, akas_len = 13205098, 38592396

progress(40)

actor_adjlist = [Node(i) for i in range(name_len)]

progress(70)

movie_adjlist = [Node(i) for i in range(akas_len)]

progress(100)

def actor_node(i, prev):
	return TraverseNode(i, prev, True)

def movie_node(i, prev):
	return TraverseNode(i, prev, False)

def search(first):
	if first: print("Enter the name of an actor, or 'exit' to exit:")
	
	actor = input("> ")
	
	if actor == "exit": return
	
	print(f"Searching for {actor} in namelist")
	
	actor_id = -1
	
	for i in range(name_len):
		if actor_namelist[i] == actor:
			actor_id = i
			break
		progress(i*100/name_len)
	
	if actor_id == -1:
		print(f"\nCouldn't find {actor} in namelist\n")
		search(False)
		return
		
	else:
		print(f"\nFound {actor} at index {actor_id}")
	
	print("Building search queue")
	
	search_queue = []
	front_index = 0
	
	actors_visited, movies_visited = set(), set()
	
	search_queue.append(TraverseNode(actor_id, None, True))
	actors_visited.add(actor_id)
	
	print("Moving through graph")
	
	iterations, depth = 0, 0
	front = search_queue[0]
	
	found_kevin_bacon = False
	if actor == "Kevin Bacon": found_kevin_bacon = True
	
	while not found_kevin_bacon and front_index < len(search_queue):
		front = search_queue[front_index]
		front_index += 1
		
		finished = False
		
		if front.is_act:
			if actor_adjlist[front.id].init:
				for movie_id in actor_adjlist[front.id]:
					if not movie_id in movies_visited:
						movies_visited.add(movie_id)
						search_queue.append(
							movie_node(actor_adjlist[front.id], front)
						)
		
		else:
			if movie_adjlist[front.id].init:
				for actor_id in movie_adjlist[front.id]:
					if not actor_id in actors_visited:
						actors_visited.add(actor_id)
						search_queue.append(
							actor_node(movie_adjlist[front.id], front)
						)
						
						if actor_id == 101:
							front = search_queue[-1]
							found_kevin_bacon = True
							finished = True
							break
		
		if finished: break
		
		if front.depth > depth:
			print(f"Depth {front.depth} ({front_index} iterations)")
			depth = front.depth
	
	if not found_kevin_bacon:
		print(f"\nCouldn't find a path from {actor} to Kevin Bacon.\n")
		search(False)
		return
	
	print("\nFound Kevin Bacon!")
	
	path = []
	
	while front.prev:
		path.append(front)
		
		front = front.prev
	
	print("Path:")
	
	print(actor, end="")
	
	while len(path) > 0:
		node = path.pop()
		
		if node.is_act:
			print(actor_namelist[node.id], end="")
			if node.id == 101: break
		
		else:
			print(f" --- ({movie_namelist[node.id]}) --> ", end="")
	
	print("\n")
	
	search(True)

print("Opening files")

act_name = open("actors_name.list").readlines()
act_adj = open("actors_adj.list").readlines()
mov_name = open("movies_name.list").readlines()
mov_adj = open("movies_adj.list").readlines()

progress(100)

print("Building adjacency lists")

for i in range(name_len):
	actor_adjlist[i].fill([int(i) for i in act_adj[i].strip().split(" ")[1:]])
	progress(i*100/(name_len+akas_len))

for i in range(akas_len):
	movie_adjlist[i].fill([int(i) for i in mov_adj[i].strip().split(" ")[1:]])
	progress((i+name_len+1)*100/(name_len+akas_len))

progress(100)

print("Reading names")

for i in range(name_len):
	actor_namelist.append(act_name[i].strip())
	progress(i*100/(name_len+akas_len))

for i in range(akas_len):
	movie_namelist.append(mov_name[i].strip())
	progress(i*100/(name_len+akas_len))

progress(100)

print("Closing files")

act_adj.close()
act_name.close()
mov_adj.close()
mov.name.close()

progress(100)

print("Ready")

search(True)

print("Exiting")
